src/aspire/denoising/adaptive_support.py

```diff
@@ -0,0 +1,107 @@
+import logging
+
+import numpy as np
+
+from aspire.image import Image
+from aspire.numeric import fft
+from aspire.utils.coor_trans import grid_2d
+
+logger = logging.getLogger(__name__)
+
+
+def adaptive_support(images, energy_threshold=0.99):
+    """
+    Determine size of the compact support in both real and Fourier Space.
+
+    Images are scaled in real space so that the noise variance in both
+    real and Fourier domains is similar.
+
+    Returns c_limit (support in Fourier space), R_limit (support in real space).
+
+    Fourier c_limit is scaled in range [0, 0.5].
+    R_limit is in pixels [0, Image.res/2].
+
+    :param images: Input `Image` stack
+    :param energy_threshold: [0, 1] threshold limit
+    :return: (c_limit, R_limit)
+    """
+
+    # Sanity Check Input is Image stack
+    if not isinstance(images, Image):
+        raise RuntimeError("adaptive_support expects Image instance")
+
+    # Sanity Check Threshold is in range
+    if energy_threshold <= 0 or energy_threshold > 1:
+        raise ValueError(
+            f"Given energy_threshold {energy_threshold} outside sane range [0,1]"
+        )
+
+    L = images.res
+    N = L // 2
+    P = images.n_images
+
+    r = grid_2d(L, shifted=False, normalized=False, dtype=images.dtype)["r"]
+
+    # Transform to Fourier space
+    imgf = fft.centered_fft2(images.asnumpy())
+
+    # demean
+    img = images.asnumpy() * L
+    mean = np.mean(img, axis=0)
+    img -= mean
+
+    # Construct a mask of corner values, outside max radiaus N
+    corner_mask = r.flatten() > N
+    img_corner = img.reshape(P, L * L)[:, corner_mask]
+    imgf_corner = imgf.reshape(P, L * L)[:, corner_mask]
+
+    # Note that the noise theoretically same, but in practice differ
+    # Take smaller so we do not yield negative variance or PSD later
+    noise_var = min(np.var(img_corner), np.var(imgf_corner))
```

Comment:
These lines (minus the `np.var(imgf_corner)` part) essentially corresponds to applying a `WhiteNoiseEstimator`, so this should probably be refactored in the future. Right now, that only takes an `ImageSource` as input, but if this is eventually rewritten to take `ImageSource`s as input, it should be fairly straightforward.

Reply:
Yep, I replaced with `WhiteNoiseEstimator` after changing to an ImageSource.